File: tests/mobi/hsz/idea/gitignore/util/GlobTest.java
package mobi.hsz.idea.gitignore.util;

import com.intellij.openapi.vfs.VirtualFile;
import mobi.hsz.idea.gitignore.Common;
import mobi.hsz.idea.gitignore.IgnoreBundle;
import mobi.hsz.idea.gitignore.file.type.kind.GitFileType;
import mobi.hsz.idea.gitignore.psi.IgnoreEntry;
import org.junit.Assert;
import org.junit.Test;

import java.lang.reflect.InvocationTargetException;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

public class GlobTest extends Common<Glob> {

    @Test
    public void testPrivateConstructor() throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
        privateConstructor(Glob.class);
    }

    @Test
    public void testFind() {
        Glob.clearCache();
        myFixture.configureByText(GitFileType.INSTANCE, createIgnoreContent("foo.txt", "bar.txt", "buz.txt", "vcsdir", "dir"));

        final VirtualFile rootFile = getFixtureRootFile();
        final List<IgnoreEntry> children = getFixtureChildrenEntries();

        myFixture.addFileToProject("bar.txt", "bar content");
        myFixture.addFileToProject("buz.txt", "buz content");
        myFixture.addFileToProject("vcsdir/.git/info", "vcsdir git info content");
        myFixture.addFileToProject("dir/buz.txt", "buz2 content");
        myFixture.addFileToProject("dir/biz.txt", "buz2 content");

        VirtualFile dir = rootFile.findChild("dir");
        assertNotNull(dir);

        /** {@link Glob#find(VirtualFile, IgnoreEntry)} test */
        Map<IgnoreEntry, List<VirtualFile>> result = Glob.find(rootFile, children, new MatcherUtil(), false);
        List<VirtualFile> item;

        // foo.txt
        item = result.get(children.get(0));
        assertNotNull(result);
        assertEmpty(item);

        // bar.txt
        item = result.get(children.get(1));
        assertNotNull(result);
        assertNotEmpty(item);
        assertEquals(item.size(), 1);
        assertTrue(item.contains(rootFile.findChild("bar.txt")));

        // buz.txt
        item = result.get(children.get(2));
        assertNotNull(result);
        assertNotEmpty(item);
        assertEquals(item.size(), 2);
        assertTrue(item.contains(rootFile.findChild("buz.txt")));
        assertTrue(item.contains(dir.findChild("buz.txt")));

        // ignore VCS directory
        item = result.get(children.get(3));
        assertNotNull(result);
        assertNotEmpty(item);
        assertEquals(item.size(), 1);
        assertTrue(item.contains(rootFile.findChild("vcsdir")));

        // dir
        item = result.get(children.get(4));
        assertNotNull(result);
        assertNotEmpty(item);
        assertEquals(item.size(), 1);
        assertTrue(item.contains(rootFile.findChild("dir")));

        /** {@link Glob#find(VirtualFile, IgnoreEntry, boolean)} test */
        // dir not includeNested
        assertNotNull(result);
        assertNotEmpty(item);
        assertEquals(item.size(), 1);
        assertTrue(item.contains(rootFile.findChild("dir")));

        // dir includeNested
        result = Glob.find(rootFile, children, new MatcherUtil(), true);
        item = result.get(children.get(4));
        assertNotNull(result);
        assertNotEmpty(item);
        assertEquals(item.size(), 3);
        assertTrue(item.contains(rootFile.findChild("dir")));
        assertTrue(item.contains(dir.findChild("buz.txt")));
        assertTrue(item.contains(dir.findChild("biz.txt")));
    }

    @Test
    public void testFindAsPaths() {
        Glob.clearCache();
        myFixture.configureByText(GitFileType.INSTANCE, createIgnoreContent("foo.txt", "bar.txt", "buz.txt", "vcsdir", "dir"));

        final VirtualFile rootFile = getFixtureRootFile();
        final List<IgnoreEntry> children = getFixtureChildrenEntries();


        myFixture.addFileToProject("bar.txt", "bar content");
        myFixture.addFileToProject("buz.txt", "buz content");
        myFixture.addFileToProject("vcsdir/.git/info", "vcsdir git info content");
        myFixture.addFileToProject("dir/buz.txt", "buz2 content");
        myFixture.addFileToProject("dir/biz.txt", "buz2 content");

        final VirtualFile dir = rootFile.findChild("dir");
        assertNotNull(dir);

        /** {@link Glob#findAsPaths(VirtualFile, IgnoreEntry)} test */
        Map<IgnoreEntry, Set<String>> result = Glob.findAsPaths(rootFile, children, new MatcherUtil(), false);
        Set<String> item;

        // foo.txt
        item = result.get(children.get(0));
        assertNotNull(result);
        assertEmpty(item);

        // bar.txt
        item = result.get(children.get(1));
        assertNotNull(result);
        assertNotEmpty(item);
        assertEquals(item.size(), 1);
        assertTrue(item.contains("bar.txt"));

        // buz.txt
        item = result.get(children.get(2));
        assertNotNull(result);
        assertNotEmpty(item);
        assertEquals(item.size(), 2);
        assertTrue(item.contains("buz.txt"));
        assertTrue(item.contains("dir/buz.txt"));
    }

    @Test
    public void testCreatePattern() throws Exception {
        Glob.clearCache();
        Pattern pattern;

        pattern = Glob.createPattern("file.txt", IgnoreBundle.Syntax.GLOB);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("file.txt").matches());
        assertTrue(pattern.matcher("dir/file.txt").matches());
        assertTrue(pattern.matcher("dir/subdir/file.txt").matches());
        Assert.assertFalse(pattern.matcher("file1.txt").matches());
        Assert.assertFalse(pattern.matcher("otherfile.txt").matches());

        pattern = Glob.createPattern("file*.txt", IgnoreBundle.Syntax.GLOB);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("file.txt").matches());
        assertTrue(pattern.matcher("dir/file.txt").matches());

        pattern = Glob.createPattern("fil[eE].txt", IgnoreBundle.Syntax.GLOB);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("file.txt").matches());
        assertTrue(pattern.matcher("filE.txt").matches());
        Assert.assertFalse(pattern.matcher("fild.txt").matches());

        pattern = Glob.createPattern("dir/file.txt", IgnoreBundle.Syntax.GLOB);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("dir/file.txt").matches());
        Assert.assertFalse(pattern.matcher("xdir/dir/file.txt").matches());
        Assert.assertFalse(pattern.matcher("xdir/file.txt").matches());

        pattern = Glob.createPattern("/file.txt", IgnoreBundle.Syntax.GLOB);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("file.txt").matches());
        Assert.assertFalse(pattern.matcher("dir/file.txt").matches());

        pattern = Glob.createPattern("fi**le.txt", IgnoreBundle.Syntax.GLOB);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("file.txt").matches());
        assertTrue(pattern.matcher("fi-foo-le.txt").matches());
        Assert.assertFalse(pattern.matcher("fi/le.txt").matches());
        Assert.assertFalse(pattern.matcher("fi/foo/le.txt").matches());

        pattern = Glob.createPattern("**/dir/file.txt", IgnoreBundle.Syntax.GLOB);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("foo/dir/file.txt").matches());
        assertTrue(pattern.matcher("dir/file.txt").matches());

        pattern = Glob.createPattern("/dir/**/file.txt", IgnoreBundle.Syntax.GLOB);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("dir/subdir/file.txt").matches());
        assertTrue(pattern.matcher("dir/subdir/foo/file.txt").matches());
        assertTrue(pattern.matcher("dir/file.txt").matches());

        pattern = Glob.createPattern("dir/*", IgnoreBundle.Syntax.GLOB, true);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("dir/file.txt").matches());
        assertTrue(pattern.matcher("dir/subdir/").matches());
        Assert.assertFalse(pattern.matcher("dir/").matches());

        pattern = Glob.createPattern("subdir", IgnoreBundle.Syntax.GLOB, true);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("dir/subdir/file.txt").matches());
        assertTrue(pattern.matcher("dir/subdir/").matches());
        Assert.assertFalse(pattern.matcher("dir/foo/bar.txt").matches());

        pattern = Glob.createPattern("subdir/", IgnoreBundle.Syntax.GLOB, true);
        assertNotNull(pattern);
        assertTrue(pattern.matcher("dir/subdir/file.txt").matches());
        assertTrue(pattern.matcher("dir/subdir/").matches());
        Assert.assertFalse(pattern.matcher("dir/foo/bar.txt").matches());
    }

}

File: src/test/kotlin/mobi/hsz/idea/gitignore/util/GlobTest.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package mobi.hsz.idea.gitignore.util

import com.intellij.openapi.components.service
import com.intellij.testFramework.UsefulTestCase
import junit.framework.TestCase
import mobi.hsz.idea.gitignore.Common
import mobi.hsz.idea.gitignore.IgnoreBundle
import mobi.hsz.idea.gitignore.file.type.IgnoreFileType
import mobi.hsz.idea.gitignore.services.IgnoreMatcher
import org.junit.Assert
import org.junit.Test
import java.lang.reflect.InvocationTargetException

@Suppress("UnsafeCallOnNullableType")
class GlobTest : Common<Glob>() {

    @Test
    @Throws(
        InvocationTargetException::class,
        NoSuchMethodException::class,
        InstantiationException::class,
        IllegalAccessException::class
    )
    fun testPrivateConstructor() {
        privateConstructor(Glob::class.java)
    }

    @Test
    fun testFind() {
        myFixture.apply {
            configureByText(
                IgnoreFileType.INSTANCE,
                createIgnoreContent("foo.txt", "bar.txt", "buz.txt", "vcsdir", "dir")
            )
            addFileToProject("bar.txt", "bar content")
            addFileToProject("buz.txt", "buz content")
            addFileToProject("vcsdir/.git/info", "vcsdir git info content")
            addFileToProject("dir/buz.txt", "buz2 content")
            addFileToProject("dir/biz.txt", "buz2 content")
        }
        val dir = fixtureRootFile.findChild("dir")
        TestCase.assertNotNull(dir)

        val matcher = project.service<IgnoreMatcher>()
        var result = Glob.find(fixtureRootFile, fixtureChildrenEntries, matcher, false)

        // foo.txt
        result[fixtureChildrenEntries[0]]!!.let {
            TestCase.assertNotNull(result)
            UsefulTestCase.assertEmpty(it)
        }

        // bar.txt
        result[fixtureChildrenEntries[1]]!!.let {
            TestCase.assertNotNull(result)
            UsefulTestCase.assertNotEmpty(it)
            TestCase.assertEquals(it.size, 1)
            TestCase.assertTrue(it.contains(fixtureRootFile.findChild("bar.txt")))
        }

        // buz.txt
        result[fixtureChildrenEntries[2]]!!.let {
            TestCase.assertNotNull(result)
            UsefulTestCase.assertNotEmpty(it)
            TestCase.assertEquals(it.size, 2)
            TestCase.assertTrue(it.contains(fixtureRootFile.findChild("buz.txt")))
            TestCase.assertTrue(it.contains(dir!!.findChild("buz.txt")))
        }

        // ignore VCS directory
        result[fixtureChildrenEntries[3]]!!.let {
            TestCase.assertNotNull(result)
            UsefulTestCase.assertNotEmpty(it)
            TestCase.assertEquals(it.size, 1)
            TestCase.assertTrue(it.contains(fixtureRootFile.findChild("vcsdir")))
        }

        // dir
        result[fixtureChildrenEntries[4]]!!.let {
            TestCase.assertNotNull(result)
            UsefulTestCase.assertNotEmpty(it)
            TestCase.assertEquals(it.size, 1)
            TestCase.assertTrue(it.contains(fixtureRootFile.findChild("dir")))

            // dir not includeNested
            TestCase.assertNotNull(result)
            UsefulTestCase.assertNotEmpty(it)
            TestCase.assertEquals(it.size, 1)
            TestCase.assertTrue(it.contains(fixtureRootFile.findChild("dir")))
        }

        // dir includeNested
        result = Glob.find(fixtureRootFile, fixtureChildrenEntries, matcher, true)
        result[fixtureChildrenEntries[4]]!!.let {
            TestCase.assertNotNull(result)
            UsefulTestCase.assertNotEmpty(it)
            TestCase.assertEquals(it.size, 3)
            TestCase.assertTrue(it.contains(fixtureRootFile.findChild("dir")))
            TestCase.assertTrue(it.contains(dir?.findChild("buz.txt")))
            TestCase.assertTrue(it.contains(dir?.findChild("biz.txt")))
        }
    }

    @Test
    fun testFindAsPaths() {
        myFixture.apply {
            configureByText(
                IgnoreFileType.INSTANCE,
                createIgnoreContent("foo.txt", "bar.txt", "buz.txt", "vcsdir", "dir")
            )
            addFileToProject("bar.txt", "bar content")
            addFileToProject("buz.txt", "buz content")
            addFileToProject("vcsdir/.git/info", "vcsdir git info content")
            addFileToProject("dir/buz.txt", "buz2 content")
            addFileToProject("dir/biz.txt", "buz2 content")
        }
        val dir = fixtureRootFile.findChild("dir")
        TestCase.assertNotNull(dir)

        val matcher = project.service<IgnoreMatcher>()
        val result = Glob.findAsPaths(fixtureRootFile, fixtureChildrenEntries, matcher, false)

        // foo.txt
        result[fixtureChildrenEntries[0]]!!.let {
            TestCase.assertNotNull(result)
            UsefulTestCase.assertEmpty(it)
        }

        // bar.txt
        result[fixtureChildrenEntries[1]]!!.let {
            TestCase.assertNotNull(result)
            UsefulTestCase.assertNotEmpty(it)
            TestCase.assertEquals(it.size, 1)
            TestCase.assertTrue(it.contains("bar.txt"))
        }

        // buz.txt
        result[fixtureChildrenEntries[2]]!!.let {
            TestCase.assertNotNull(result)
            UsefulTestCase.assertNotEmpty(it)
            TestCase.assertEquals(it.size, 2)
            TestCase.assertTrue(it.contains("buz.txt"))
            TestCase.assertTrue(it.contains("dir/buz.txt"))
        }
    }

    @Test
    @Throws(Exception::class)
    @Suppress("LongMethod")
    fun testCreatePattern() {
        Glob.createPattern("file.txt", IgnoreBundle.Syntax.GLOB).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("file.txt").matches())
            TestCase.assertTrue(it.matcher("dir/file.txt").matches())
            TestCase.assertTrue(it.matcher("dir/subdir/file.txt").matches())
            Assert.assertFalse(it.matcher("file1.txt").matches())
            Assert.assertFalse(it.matcher("otherfile.txt").matches())
        }

        Glob.createPattern("file*.txt", IgnoreBundle.Syntax.GLOB).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("file.txt").matches())
            TestCase.assertTrue(it.matcher("dir/file.txt").matches())
        }

        Glob.createPattern("fil[eE].txt", IgnoreBundle.Syntax.GLOB).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("file.txt").matches())
            TestCase.assertTrue(it.matcher("filE.txt").matches())
            Assert.assertFalse(it.matcher("fild.txt").matches())
        }

        Glob.createPattern("dir/file.txt", IgnoreBundle.Syntax.GLOB).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("dir/file.txt").matches())
            Assert.assertFalse(it.matcher("xdir/dir/file.txt").matches())
            Assert.assertFalse(it.matcher("xdir/file.txt").matches())
        }

        Glob.createPattern("/file.txt", IgnoreBundle.Syntax.GLOB).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("file.txt").matches())
            Assert.assertFalse(it.matcher("dir/file.txt").matches())
        }

        Glob.createPattern("fi**le.txt", IgnoreBundle.Syntax.GLOB).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("file.txt").matches())
            TestCase.assertTrue(it.matcher("fi-foo-le.txt").matches())
            Assert.assertFalse(it.matcher("fi/le.txt").matches())
            Assert.assertFalse(it.matcher("fi/foo/le.txt").matches())
        }

        Glob.createPattern("**/dir/file.txt", IgnoreBundle.Syntax.GLOB).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("foo/dir/file.txt").matches())
            TestCase.assertTrue(it.matcher("dir/file.txt").matches())
        }

        Glob.createPattern("/dir/**/file.txt", IgnoreBundle.Syntax.GLOB).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("dir/subdir/file.txt").matches())
            TestCase.assertTrue(it.matcher("dir/subdir/foo/file.txt").matches())
            TestCase.assertTrue(it.matcher("dir/file.txt").matches())
        }

        Glob.createPattern("dir/*", IgnoreBundle.Syntax.GLOB, true).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("dir/file.txt").matches())
            TestCase.assertTrue(it.matcher("dir/subdir/").matches())
            Assert.assertFalse(it.matcher("dir/").matches())
        }

        Glob.createPattern("subdir", IgnoreBundle.Syntax.GLOB, true).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("dir/subdir/file.txt").matches())
            TestCase.assertTrue(it.matcher("dir/subdir/").matches())
            Assert.assertFalse(it.matcher("dir/foo/bar.txt").matches())
        }

        Glob.createPattern("subdir/", IgnoreBundle.Syntax.GLOB, true).let {
            TestCase.assertNotNull(it)
            TestCase.assertTrue(it!!.matcher("dir/subdir/file.txt").matches())
            TestCase.assertTrue(it.matcher("dir/subdir/").matches())
            Assert.assertFalse(it.matcher("dir/foo/bar.txt").matches())
        }
    }
}
